### PR TITLE
[2d] Smooth Falloff for 2D Shape Lights.

### PR DESCRIPTION
**Pull request by venkify**
Please note :
  . All concave/convex polygons handled properly for both Shape and Offsets.

Future Todo:
  . Handle Offset overlaps and 

# **Please read;**
## **PR Workflow for the Graphics repository:**
* **All PRs must be opened as draft initially**
* Reviewers can be added while the PR is still in draft
* The PR can be marked as “Ready for Review” once the reviewers have confirmed that **no more changes are needed**
* Tests will start automatically after the PR is marked as “Ready for Review”
* **Do not use [skip ci]** - this can break some of our tooling
* Read the [Graphics repository & Yamato FAQ](http://go/graphics-yamato-faq).

### Checklist for PR maker
- [ ] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [ ] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [ ]  Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR. If you do add documentation, make sure to add the relevant Graphics Docs team member as a reviewer of the PR. If you are not sure which person to add, see the [Docs team contacts sheet](https://docs.google.com/spreadsheets/d/1rgUWWgwLFEHIQ3Rz-LnK6PAKmbM49DZZ9al4hvnztOo/edit#gid=1058860420).
- [ ] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
Why is this PR needed, what hard problem is it solving/fixing?

---
### Testing status

**Manual Tests**: What did you do?
- [ ] Opened test project + Run graphic tests locally
- [ ] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo + Prefab overrides + Alignment in Preset
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in developer mode, you have a button at end of resources that check the paths)
- Other: 

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics

Any test projects to go with this to help reviewers?

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

**Review by unity-cchu — APPROVED** [review]

File: com.unity.render-pipelines.universal/Runtime/2D/Light2DAuthoring.cs [hunk 1]
```diff
@@ -26,11 +26,48 @@ void Reset()
         {
             m_ShapePath = new Vector3[] { new Vector3(-0.5f, -0.5f), new Vector3(0.5f, -0.5f), new Vector3(0.5f, 0.5f), new Vector3(-0.5f, 0.5f) };
         }
+        
+        // Rough shape only used in Inspector for quick preview. 
+        static List<Vector2> GetFalloffShape(Vector3[] shapePath)
```
Comment (rainsing) [inline com.unity.render-pipelines.universal/Runtime/2D/Light2DAuthoring.cs:31]: I know you only moved this method, but we should take this chance to rename it to what it actually does: GetExtrusionDir().

File: com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs [hunk 2]
```diff
@@ -170,127 +171,250 @@ public static Bounds GenerateSpriteMesh(Mesh mesh, Sprite sprite)
             mesh.SetVertexBufferParams(vertices.Length, SpriteLightMeshVertex.VertexLayout);
             mesh.SetVertexBufferData(vertices, 0, 0, vertices.Length);
             mesh.SetIndices(srcIndices, MeshTopology.Triangles, 0, true);
-            return mesh.GetSubMesh(0).bounds;
+            return mesh.bounds;
```
Comment (rainsing) [inline com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs:174]: Are you sure about this change? According to my previous testing, the last boolean parameter of `Mesh.SetIndices()` only tells the mesh to recalculate the bounds for the submesh in question, not for the whole mesh. Returning Mesh.bounds here was causing culling issues in Lost Crypt.

File: com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs [hunk 3]
```diff
@@ -170,127 +171,250 @@ public static Bounds GenerateSpriteMesh(Mesh mesh, Sprite sprite)
...
+                extrusionDir[i] = dir;
+            }
+            
+            for (var i = 0; i < shapePathLength; i++)
+            {
+                var j = (i == shapePath.Length - 1) ? 0 : (i + 1);
+                var sPI = shapePath[i];
+                var sPJ = shapePath[j];
+                var eDI = extrusionDir[i];
+                var eDJ = extrusionDir[i];
```
Comment (rainsing) [inline com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs:232]: Is this intentional? eD**J** = extrusionDir[**i**];

I tried changing it to eDJ = extrusionDir[j], but the result was wrong.

File: com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs [hunk 4]
```diff
@@ -170,127 +171,250 @@ public static Bounds GenerateSpriteMesh(Mesh mesh, Sprite sprite)
...
             }
+        }
 
-            var indexOffset = 0;
-            foreach (var v in indicesI)
+        static void GenerateBevels(NativeArray<float4> bevel, float fallOff, Color meshInteriorColor, NativeArray<ushort> indices,
+            NativeArray<ParametricLightMeshVertex> vertices, ref int vcount, ref int icount)
+        {
+            var vertexCount = bevel.Length;
+            var quality = 4;
```
Comment (rainsing) [inline com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs:321]: Should we make this configurable?

File: com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs [hunk 5]
```diff
@@ -170,127 +171,250 @@ public static Bounds GenerateSpriteMesh(Mesh mesh, Sprite sprite)
...
+            var arcPoints = new NativeArray<float3>(quality + 1, Allocator.Temp);
+            for (var i = 0; i < vertexCount; i = i + 3)
             {
-                finalIndices[indexOffset++] = (ushort)v;
+                if ( 0 != bevel[i + 1].w )
+                    continue;
+                
+                // Gnerate Pivot Position
```
Comment (rainsing) [inline com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs:328]: Typo.

**Review by rainsing — CHANGES_REQUESTED** [review]
Overall nice changes! The new shapes look much better. 

Please double check the changes around mesh bounds. Unless there was a recent trunk change to the mesh API that I'm not aware of, otherwise the new code will cause culling problems.

File: com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs (hunk 2 above)
Comment (venkify) [reply com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs:174]: I think you are right. Let me check and reset it back if its the case.

File: com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs (hunk 3 above)
Comment (venkify) [reply com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs:232]: eDJ = extrusionDir[i]; is intentional.. Just was testing different values.

File: com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs (hunk 4 above)
Comment (venkify) [reply com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs:321]: Yes, let's discuss. @UnityPaul wanted minimal changes, so not yet exposed.

File: com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs (hunk 5 above)
Comment (venkify) [reply com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs:328]: Will fix.

File: com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs (hunk 2 above)
Comment (venkify) [reply com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs:376]: mesh.bounds encapsulates bounds of all submeshes. We can also get bounds for each Submesh too. For now I have changed as you suggested.

I guess once we have Transparent/Opaque Split we may need to use mesh.bounds.

File: com.unity.render-pipelines.universal/Runtime/2D/Light2DAuthoring.cs (hunk 1 above)
Comment (venkify) [reply com.unity.render-pipelines.universal/Runtime/2D/Light2DAuthoring.cs]: Renamed.

File: com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs (hunk 4 above)
Comment (venkify) [reply com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs]: The implementation has changed. Although we can choose the quality levels, let's keep default for now. Can change later.

File: com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs (hunk 5 above)
Comment (venkify) [reply com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs]: Fixed.

File: com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs (hunk 3 above)
Comment (venkify) [reply com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs]: We have a better implementation now based on Clipper. The above is not needed.

File: com.unity.render-pipelines.universal/Runtime/2D/Clipper.cs [hunk 6]
```diff
@@ -0,0 +1,5025 @@
...
+
+        //------------------------------------------------------------------------------
+    }
+
+    class ClipperException : Exception
+    {
+        public ClipperException(string description) : base(description) {}
+    }
+    //------------------------------------------------------------------------------
+} //end ClipperLib namespace
```
Comment (rainsing) [inline com.unity.render-pipelines.universal/Runtime/2D/Clipper.cs:5025]: A few questions regarding this 3rd party source:
* Do we need the review of legal?
* It introduces public APIs under the namespace UnityEngine.Experimental.Rendering.Universal, which I don't think we should do.

File: com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs [hunk 7]
```diff
@@ -57,6 +67,199 @@ private struct SpriteLightMeshVertex
...
+            tess.AddContour(inner, ContourOrientation.CounterClockwise);
+            Tessellate(tess, ElementType.Polygons, indices, vertices, meshInteriorColor, ref vcount, ref icount);
+
+            // Create falloff geometry
+            List<IntPoint> path = new List<IntPoint>();
+            var pointNoise = new System.Random();
+            for (var i = 0; i < innerShapeVertexCount; ++i)
+            {
+                var pnx = 0;// pointNoise.Next(10);
+                var pny = 0;// pointNoise.Next(10);
```
Comment (rainsing) [inline com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs:185]: Can we remove these?

**Review by rainsing — CHANGES_REQUESTED** [review]
We should not introduce new public APIs via the addition of the Clipper library.

File: com.unity.render-pipelines.universal/Runtime/2D/Clipper.cs (hunk 6 above)
Comment (venkify) [reply com.unity.render-pipelines.universal/Runtime/2D/Clipper.cs:5025]: Good catch. Will remove all public APIs to internal. Not sure regarding license. I think we already use it in Unity/Trunk.

File: com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs (hunk 7 above)
Comment (venkify) [reply com.unity.render-pipelines.universal/Runtime/2D/LightUtility.cs]: They were used for testing. Yes, these have been removed in the latest.

File: com.unity.render-pipelines.universal/Runtime/2D/Clipper.cs (hunk 6 above)
Comment (venkify) [reply com.unity.render-pipelines.universal/Runtime/2D/Clipper.cs:5025]: Fixed in https://github.com/Unity-Technologies/Graphics/pull/1707/commits/57dfa2a4af15ad19146d1cfb14fc4454719af1ca

**Review by rainsing — APPROVED** [review]

**Comment by rainsing** [comment]
@venkify Just follow the PR guideline and fill the relevant information. When you hit that Ready for review button CI will automatically trigger. When everything is green add the read-to-merge label.

**Review by UnityPaul — APPROVED** [review]

**Review by darrenquek — APPROVED** [review]
Tested and verified on various projects like LOSTCRYPT and also tested some edge case shapes.
